File: FrontEnd/src/services/appointment.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpHeaders, HttpResponse} from "@angular/common/http";
import { Observable } from "rxjs";
import {AppointmentRequest, AppointmentResponse} from '../app/Appointment/appointment.model'

@Injectable({
  providedIn: 'root'
})
export class AppointmentService {

  private baseUrl = 'http://localhost:8082/appointment';

  private httpOptions = {
    headers: new HttpHeaders({
      'Content-Type': 'application/json'
    })
  };


  constructor(private http: HttpClient) {}

// Create Appointment
  createAppointment(appointment: {
    familyId: number;
    doctorId: number;
    date: Date;
    description: string;
    emergency: number
  }): Observable<AppointmentResponse> {
    return this.http.post<AppointmentResponse>(`${this.baseUrl}/create`, appointment, this.httpOptions);
  }

  // Get All Appointments
  getAllAppointments(): Observable<AppointmentResponse[]> {
    return this.http.get<AppointmentResponse[]>(`${this.baseUrl}/appointment`);
  }

  // Get Appointment by ID
  getAppointmentById(id: number): Observable<AppointmentResponse> {
    return this.http.get<AppointmentResponse>(`${this.baseUrl}/appointment/${id}`);
  }

  // Update Appointment
  updateAppointment(id: number, appointment: AppointmentRequest): Observable<AppointmentResponse> {
    return this.http.put<AppointmentResponse>(`${this.baseUrl}/appointment/${id}`, appointment, this.httpOptions);
  }

  // Delete Appointment
  deleteAppointment(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/appointment/${id}`);
  }

  // Remove Doctor from Appointment (sets Appointment's doctor_id to NULL directly)
  removeDoctorFromAppointment(appointmentId: number): Observable<AppointmentResponse> {
    return this.http.patch<AppointmentResponse>(`${this.baseUrl}/appointment/${appointmentId}/removeDoctor`, null);
  }

}

File: BackEnd/src/main/java/com/example/care4elders/services/AppointmentService.java
package com.example.care4elders.services;

import com.example.care4elders.controllers.DTO.AppointmentRequest;
import com.example.care4elders.controllers.DTO.AppointmentResponse;
import com.example.care4elders.model.Appointment;

import java.util.List;


public interface AppointmentService {
    // Helper method to convert Entity to Response DTO
    AppointmentResponse convertToDto(Appointment appointment);


    // CREATE
    AppointmentResponse createAppointment(AppointmentRequest request);

    // READ (by ID)
    AppointmentResponse getAppointmentById(Long id);

    // READ (all)
    List<AppointmentResponse> getAllAppointments();

    // UPDATE
    AppointmentResponse updateAppointment(Long id, AppointmentRequest request);

    // DELETE
    void deleteAppointment(Long id);

    List<AppointmentResponse> getAppointmentsByDoctorId(Long doctorId);

    List<AppointmentResponse> getAppointmentsByFamilyId(Long familyId);
}
